Ros2 devel

Karsten1987 opened the pull request:
This is a fair mix of commits started by @bponsler and @MarcTestier in order to port laser_proc to ros2. I am opening this in order to have a single upstream port rather than multiple forks.

This is a pretty straight forward migration without any chances in design or code style. 

A ros2 port of this is further needed to unblock the migration work of URG: https://github.com/ros-drivers/urg_node/pull/50

bponsler committed:
Initial port to ros 2

Initial port of the laser_proc package to ros 2

bponsler committed:
Exporting dep libs, fixed tests

bponsler committed:
Removed missing packages

bponsler committed:
Updated to use cmake 3.5

bponsler committed:
Fixing mistakes

Removing creation of private node handle which is not allowed in ROS 2.

bponsler committed:
Fixing typo in package manifest

guchaojie committed:
Fixed compile error: cannot find class_loader_register_macro.h
when ament build add --isolated option

bponsler committed:
Merge pull request #1 from guchaojie/ros2-devel

Fixed compile error: cannot find class_loader_register_macro.h when ament build add --isolated option

Karsten1987 committed:
ros2 ardent compliant

bponsler committed:
Merge pull request #2 from Karsten1987/ros2-devel

ros2 ardent compliant

MarcTestier committed:
Minor update to ROS 2 Bouncy

bponsler committed:
Merge pull request #3 from MarcTestier/ros2-devel

Minor update to ROS 2 Bouncy

Karsten1987 committed:
avoid deprecation warning

Signed-off-by: Karsten Knese <karsten@openrobotics.org>

Karsten1987 committed:
use publish overload function

Signed-off-by: Karsten Knese <karsten@openrobotics.org>

Karsten1987 committed:
some cleanups

Signed-off-by: Karsten Knese <karsten@openrobotics.org>

Karsten1987 committed:
Merge branch 'ros2-devel' into ros2-devel

Karsten1987 committed:
enable linters

Karsten1987 commented:
@chadrockey Do you mind having a look at this? Ideally, we would have this landed upstream and released in dashing/eloquent to push forward a release of the urg laser driver.

Karsten1987 committed:
address warning

clalancette commented on CMakeLists.txt:
```diff
@@ -1,88 +1,87 @@
-target_link_libraries(LaserProcNodelet laser_proc_ROS ${catkin_LIBRARIES})
+add_library(laser_proc_nodelet src/laser_proc_nodelet.cpp)
+target_link_libraries(laser_proc_nodelet laser_proc)
+ament_target_dependencies(laser_proc_nodelet
```
I think we should get away from the `nodelet` terminology here, as it is a source of confusion.  In ROS 2, we are typically using the term "component", so I'd suggest that (or something like it) instead.

Karsten1987 replied:
I know too little about how laser proc is being used to judge whether we need this in the first place. What does laser proc do when started as a node?

clalancette replied:
I talked about this with @Karsten1987 offline, and it appears that laser_proc subscribes to a `MultiEchoLaserScan` message.  When it gets one, it republishes that data on a different topic, and also picks out the first, last, and most intense scan and publishes them as `LaserScan` messages.

So with all of that being said, I think it makes sense to have this be available as both a regular node and a composable node.  I'm not saying that we necessarily need to do all of that work in this PR, but at least getting away from the nodelet terminology here makes sense to me.

clalancette commented on include/laser_proc/laser_transport.hpp:
```diff
@@ -27,31 +27,29 @@
+    rclcpp::Node::SharedPtr & nh, uint32_t queue_size,
+    /* const ros::SubscriberStatusCallback& connect_cb,
+    const ros::SubscriberStatusCallback& disconnect_cb = ros::SubscriberStatusCallback(),
+    const ros::VoidPtr& tracked_object = ros::VoidPtr(), bool latch = false, */
```
Can we just remove this commented-out code?

clalancette commented on src/laser_proc.cpp:
```diff
@@ -27,20 +27,146 @@
+
+#include "laser_proc/laser_proc.hpp"
+
+using namespace laser_proc;  // NOLINT
```
Instead of doing this with a `NOLINT`, can we instead just do:

```
namespace laser_proc
{
```

At the top here, and close out the namespace at the bottom of the file?

clalancette replied:
This comment is still open.

clalancette commented on src/laser_proc.cpp:
```diff
@@ -27,20 +27,146 @@
+    std::stringstream ss;
+    ss <<
+      "getMostIntenseScan::Size of ranges does not equal size of intensities, cannot create scan.";
+    throw std::runtime_error(ss.str());
```
Throwing an exception here means that an external process can cause us to crash just by sending a malformed message.  This should be a warning and a return rather than an exception.

clalancette replied:
This comment is still open.

Karsten1987 replied:
I was hesitating about introducing rclcpp here, but in the end opted for `RCLCPP_ERROR`. Let me know if a simple `fprintf` is preferred

clalancette commented on src/laser_proc_node.cpp:
```diff
@@ -27,24 +27,22 @@
-} //namespace laser_proc
+  rclcpp::spin(n);
 
+  return 0;
```
It's kind of superfluous, but we typically put an `rclcpp::shutdown()` at the end of main.

clalancette commented on src/laser_proc_ros.cpp:
```diff
@@ -27,31 +27,32 @@
   pub_.publish(msg);
 }
 
+/*
```
Should we just delete this commented-out code?

clalancette commented on src/laser_publisher.cpp:
```diff
@@ -0,0 +1,208 @@
+
+uint32_t LaserPublisher::getNumSubscribers() const
+{
+  // TODO(karsten1987): no way to do this in ros2 yet
```
This should be possible now; see https://github.com/ros-drivers/velodyne/blob/ec9adec6a523ab705ee526b06141aa9ad8d43c4f/velodyne_pointcloud/src/conversions/convert.cpp#L150 for an example.

clalancette reviewed (commented):
I've left some comments for improvements.

Karsten1987 commented on src/laser_proc_nodelet.cpp:
```diff
@@ -27,35 +27,36 @@
-
-}
+    lp.reset(new LaserProcROS(shared_from_this()));
+  }
```
Is this `onInit()` method a leftover of ROS1 nodelets or where is this function actually being called?

Karsten1987 committed:
cleanup API to use ROS2 components

Signed-off-by: Karsten Knese <karsten@openrobotics.org>

Karsten1987 commented:
with the latest commit I've addressed some of the review points, but more over changed quite some code which was unused and can now be handled more conveniently with the ROS2 component API.

clalancette commented on include/laser_proc/laser_proc.hpp:
```diff
@@ -27,49 +27,47 @@
-#include <sstream>
+#include "sensor_msgs/msg/laser_scan.hpp"
+#include "sensor_msgs/msg/multi_echo_laser_scan.hpp"
+#include "sensor_msgs/msg/laser_echo.hpp"
```
Nit: alphabetize.

clalancette commented on include/laser_proc/laser_publisher.hpp:
```diff
@@ -0,0 +1,101 @@
+  LaserPublisher(
+    rclcpp::Node::SharedPtr & nh, uint32_t queue_size, bool publish_echoes = true);
+
+  LaserPublisher(
```
`explicit` on both of these.

Karsten1987 replied:
can you elaborate on why?

clalancette replied:
Generally to avoid surprising type conversions.  It makes the callers have to be more type-conscious, which I think is generally a good thing.

clalancette reviewed (commented):
A few more things left open, but this is looking much better!

Karsten1987 committed:
more review

Signed-off-by: Karsten Knese <karsten@openrobotics.org>

clalancette reviewed (approved):
Thanks for iterating.  This looks good to me now.

clalancette commented:
@chadrockey Any thoughts on this PR?  It looks good to me, so unless you have any thoughts or objections, @Karsten1987 or I will merge at the end of the week.

clalancette commented:
All right, I'm going to go ahead and merge.  If there are any comments you want to add, please feel free.
